pkg/roachprod/failureinjection/failures/disk_stall.go

```diff
@@ -0,0 +1,466 @@
+	if err = s.Run(ctx, l, s.c.Nodes, `sudo tune2fs -O ^has_journal `+dev); err != nil {
+		return errors.WithHintf(err, "disabling journaling fails if the cluster has been started")
+	}
+	if err = s.Run(ctx, l, s.c.Nodes, `echo "0 $(sudo blockdev --getsz `+dev+`) linear `+dev+` 0" | `+
+		`sudo dmsetup create data1`); err != nil {
+		return err
+	}
+	// This has occasionally been seen to fail with "Device or resource busy",
+	// with no clear explanation. Try to find out who it is.
+	if err = s.Run(ctx, l, s.c.Nodes, "sudo bash -c 'ps aux; dmsetup status; mount; lsof'"); err != nil {
+		return err
+	}
+
+	if err = s.Run(ctx, l, s.c.Nodes, `sudo mount /dev/mapper/data1 /mnt/data1`); err != nil {
+		return err
+	}
+
+	if diskStallArgs.RestartNodes {
+		if err = s.StartCluster(ctx, l); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *DmsetupDiskStaller) Inject(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	nodes := args.(DiskStallArgs).Nodes
+	l.Printf("stalling disk I/O on nodes %d", nodes)
+	return s.Run(ctx, l, nodes, `sudo dmsetup suspend --noflush --nolockfs data1`)
+}
+
+func (s *DmsetupDiskStaller) Restore(
+	ctx context.Context, l *logger.Logger, args FailureArgs,
+) error {
+	diskStallArgs := args.(DiskStallArgs)
+	nodes := diskStallArgs.Nodes
+	l.Printf("unstalling disk I/O on nodes %d", nodes)
+	if err := s.Run(ctx, l, nodes, `sudo dmsetup resume data1`); err != nil {
+		return err
+	}
+	// If the disk stall was injected for long enough that the cockroach process
+	// detected it and shut down the node, then restart it.
+	if err := s.PingNode(ctx, l, nodes); err != nil && diskStallArgs.RestartNodes {
+		l.Printf("failed to connect, assuming node exited and restarting: %v", err)
+		return s.StartNodes(ctx, l, nodes)
+	}
+	return nil
+}
+
+func (s *DmsetupDiskStaller) Cleanup(
+	ctx context.Context, l *logger.Logger, args FailureArgs,
+) error {
+	diskStallArgs := args.(DiskStallArgs)
+	if diskStallArgs.RestartNodes {
+		stopOpts := roachprod.DefaultStopOpts()
+		if err := s.StopCluster(ctx, l, stopOpts); err != nil {
+			return err
+		}
+	}
+
+	dev, err := s.DiskDeviceName(ctx, l)
+	if err != nil {
+		return err
+	}
+
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo dmsetup resume data1`); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo umount /mnt/data1`); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo dmsetup remove data1`); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo tune2fs -O has_journal `+dev); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo mount /mnt/data1`); err != nil {
+		return err
+	}
+	// Reinstall snapd.
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo apt-get install -y snapd`); err != nil {
+		return err
+	}
+
+	// When we unmounted the disk in setup, the cgroups controllers may have been removed, re-add them.
```

Comment:
This feels iffy to me to have to manually add back the controllers, but no matter how hard I tried I couldn't get them to be added back short of restarting the roachprod cluster.

Reply:
> I tried I couldn't get them to be added back short of restarting the roachprod cluster.

By "restarting", you mean VMs (not systemd)? It does look error-prone. What if we miss something; what's the chance that the cluster remains fully reusable (by independent roachtest)?